### PR TITLE
feat(scalasdk): Actions API

### PR DESCRIPTION
**Pull request by octonato**

### CODE REVIEW

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala [hunk 1]
```diff
@@ -16,5 +16,22 @@
 
 package com.akkaserverless.scalasdk
 
-//FIXME (not extend from javasdk Context?)
-trait Context extends com.akkaserverless.javasdk.Context
+trait Context {
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala:19]: 👍

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala [hunk 2]
```diff
@@ -16,5 +16,22 @@
...
+
+  /** Get the service call factory for this stateful service. */
+  def serviceCallFactory: ServiceCallFactory
+
+  /**
+   * Get an Akka gRPC client for the given service name. The same client instance is shared across components in the
+   * application. The lifecycle of the client is managed by the SDK and it should not be stopped by user code.
+   *
+   * @tparam [T]
+   *   The "service" interface generated for the service by Akka gRPC
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/Context.scala:29]: I think it's like this in ScalaDoc:
```
@tparam T ...
```

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionOptions.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import com.akkaserverless.javasdk.impl.ComponentOptions
+
+object ActionOptions {
+
+  def defaults: ActionOptions = ActionOptionsImpl(Set.empty)
+
+  private[akkaserverless] final case class ActionOptionsImpl(_forwardHeaders: Set[String]) extends ActionOptions {
+
+    override def forwardHeaders(): java.util.Set[String] = _forwardHeaders.asJava
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionOptions.scala:33]: strange that we have `java.util.Set` in Scala API, but that could be a separate issue

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionProvider.scala [hunk 4]
```diff
@@ -16,13 +16,15 @@
...
+trait ActionProvider[A <: Action] {
+  def options: ActionOptions
+
+  def serviceDescriptor: Descriptors.ServiceDescriptor
 
-//FIXME possibly The Provider will not delegate to javasdk and we'll duplicate some more code
-class ActionProvider[A <: Action](private[akkaserverless] val impl: Impl[A#Impl])
+  def newHandler(context: ActionCreationContext): ActionHandler[A]
+
+  def additionalDescriptors: Array[Descriptors.FileDescriptor]
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionProvider.scala:29]: in Views and ValueEntity we have that as `immutable.Seq[Descriptors.FileDescriptor]`

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/MessageEnvelope.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * Create a message.
+   *
+   * @param payload
+   *   The payload of the message.
+   * @return
+   *   The message.
+   */
+  def apply[T](payload: T) = new MessageEnvelopeImpl[T](payload, Metadata.empty)
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/MessageEnvelope.scala:32]: always public return type in public api

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/MessageEnvelope.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+   * Create a message.
+   *
+   * @param payload
+   *   The payload of the message.
+   * @param metadata
+   *   The metadata associated with the message.
+   * @return
+   *   The message.
+   */
+  def apply[T](payload: T, metadata: Metadata) = new MessageEnvelopeImpl[T](payload, metadata)
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/MessageEnvelope.scala:44]: public return type in public api

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/ServiceCallAdapters.scala [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+private[scalasdk] case class ScalaServiceCallFactoryAdapter(javaSdkServiceCallFactory: javasdk.ServiceCallFactory)
+    extends ServiceCallFactory {
+
+  override def lookup[T](serviceName: String, methodName: String, messageType: Class[T]): ServiceCallRef[T] =
+    ScalaServiceCallRefAdapter(javaSdkServiceCallFactory.lookup(serviceName, methodName, messageType))
+
+}
+private[scalasdk] case class JavaServiceCallAdapter(scalaSdkServiceCall: ServiceCall) extends javasdk.ServiceCall {
+  override def ref(): javasdk.ServiceCallRef[_] = JavaServiceCallRefAdapter(scalaSdkServiceCall.ref)
+  override def message(): protobuf.Any = scalaSdkServiceCall.message
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/ServiceCallAdapters.scala:38]: we probably want to use ScalaPB Any for the Scala SDK, but that can be a separate issue

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/action/ActionHandler.scala [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.akkaserverless.scalasdk.action.Action
+import com.akkaserverless.scalasdk.action.ActionContext
+import com.akkaserverless.scalasdk.action.MessageEnvelope
+import com.akkaserverless.scalasdk.impl.action.ActionHandler.HandlerNotFound
+
+object ActionHandler {
+  case class HandlerNotFound(commandName: String) extends RuntimeException
```
Comment (patriknw) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/action/ActionHandler.scala:27]: For views I skipped this one, and use the javasdk HandlerNotFound. The reasons were:
1. then I didn't have to catch convert and throw
2. it's not exposed to users anyway (only in generated Handler code)

**Review by patriknw — COMMENTED** [review]
looking good, a few small things and some things for separate issues/prs

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionOptions.scala (hunk 3 above)
Comment (octonato) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/action/ActionOptions.scala:33]: yes, we need to duplicate the Options as well.

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/action/ActionEffectImpl.scala [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+import com.akkaserverless.javasdk
+import com.akkaserverless.scalasdk.Metadata
+import com.akkaserverless.scalasdk.ServiceCall
+import com.akkaserverless.scalasdk.SideEffect
+import com.akkaserverless.scalasdk.action.Action
+import com.akkaserverless.scalasdk.impl.JavaServiceCallAdapter
+import com.akkaserverless.scalasdk.impl.JavaSideEffectAdapter
+
+object ActionEffectImpl {
```
Comment (RayRoestenburg) [inline sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/action/ActionEffectImpl.scala:32]: I wonder if you can't just delegate to java impl, like in `scalasdk.impl.ValueEntityEffectImpl` (not a big deal, but might mean less duplication)

Comment (octonato) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/action/ActionEffectImpl.scala:32]: We can explore this eventually on another PR. I just realised that Value Entities and ES Entities are encoded quite different than Actions. 

The `Impl` class is a Builder and an Effect. This is not how the Actions are implemented though. There are different `Effect` implementations.

File: sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/ServiceCallAdapters.scala (hunk 7 above)
Comment (octonato) [reply sdk/scala-sdk/src/main/scala/com/akkaserverless/scalasdk/impl/ServiceCallAdapters.scala:38]: I will change it on another PR.

**Comment by octonato** [comment]
@patriknw / @RayRoestenburg, applied code review suggestions on e1cca79. 

I will add the other requested changes on separated PRs so we can move with this one.

**Pull request merged**